### PR TITLE
Add Sponge Getter support

### PR DESCRIPTION
**Pull request by RedNesto**
This PR adds the following features (requested in the two first bullets of #329):
- annotation's target method completion
- inspection ensuring parameter type matches targeted type (includes Optional<> handling)
- (bonus) suppression of "Optional should not be used as parameter"

Not sure if some parts of the code are really clear, I'd like to know if everything is understandable enough.

### CODE REVIEW

File: src/main/kotlin/com/demonwav/mcdev/platform/sponge/completion/SpongeGetterFilterCompletionContributor.kt [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import com.demonwav.mcdev.platform.sponge.inspection.SpongeInvalidGetterTargetInspection
+import com.demonwav.mcdev.platform.sponge.util.SpongeConstants
+import com.demonwav.mcdev.platform.sponge.util.isValidSpongeListener
+import com.demonwav.mcdev.util.findContainingMethod
+import com.demonwav.mcdev.util.packageName
+import com.intellij.codeInsight.completion.*
+import com.intellij.codeInsight.lookup.AutoCompletionPolicy
+import com.intellij.codeInsight.lookup.LookupElementBuilder
+import com.intellij.lang.jvm.types.JvmReferenceType
+import com.intellij.psi.*
```
Comment (DenWav) [inline src/main/kotlin/com/demonwav/mcdev/platform/sponge/completion/SpongeGetterFilterCompletionContributor.kt]: Please use individual imports.

File: src/main/kotlin/com/demonwav/mcdev/platform/sponge/inspection/suppress/SpongeGetterParamOptionalInspectionSuppressor.kt [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        val param = element.parentOfType(PsiParameter::class) ?: return false
+        val method = param.findContainingMethod() ?: return false
+        if (!method.isValidSpongeListener()) {
+            return false
+        }
+
+        val getterAnnotation = param.getAnnotation(SpongeConstants.GETTER_ANNOTATION) ?: return false
+        val targetMethod = getterAnnotation.resolveSpongeGetterTarget() ?: return false
+
+        return targetMethod.returnType == param.type && ((param.type as JvmReferenceType).resolve()!! as PsiClass).qualifiedName == SpongeConstants.OPTIONAL
```
Comment (DenWav) [inline src/main/kotlin/com/demonwav/mcdev/platform/sponge/inspection/suppress/SpongeGetterParamOptionalInspectionSuppressor.kt]: Any `resolve` operation isn't a null-safe, please handle this safely.

File: src/main/kotlin/com/demonwav/mcdev/platform/sponge/completion/SpongeGetterFilterCompletionContributor.kt (hunk 1 above)
Comment (DemonWav) [inline src/main/kotlin/com/demonwav/mcdev/platform/sponge/completion/SpongeGetterFilterCompletionContributor.kt]: Please use individual imports.

File: src/main/kotlin/com/demonwav/mcdev/platform/sponge/inspection/SpongeInvalidGetterTargetInspection.kt [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        companion object {
+
+            fun doFix(project: Project, annotation: PsiAnnotation) {
+                val value = JavaPsiFacade.getElementFactory(project).createExpressionFromText("\"\"", annotation)
+                val newValue = annotation.setDeclaredAttributeValue(null, value)
+                val editor = PsiEditorUtil.Service.getInstance().findEditorByPsiElement(annotation) ?: return
+                editor.caretModel.removeSecondaryCarets()
+                editor.selectionModel.removeSelection()
+                editor.caretModel.moveToOffset(newValue.textOffset + 1)
+                thread {
```
Comment (DenWav) [inline src/main/kotlin/com/demonwav/mcdev/platform/sponge/inspection/SpongeInvalidGetterTargetInspection.kt]: What are you trying to do here? Calling `invokeLater` immediately in a new `thread` doesn't really make any sense.

**Review by DenWav — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/demonwav/mcdev/platform/sponge/completion/SpongeGetterFilterCompletionContributor.kt (hunk 1 above)
Comment (RedNesto) [reply src/main/kotlin/com/demonwav/mcdev/platform/sponge/completion/SpongeGetterFilterCompletionContributor.kt]: Oh I didn't notice the `intellij-style.xml` file. I'll reformat the files with this profile.

File: src/main/kotlin/com/demonwav/mcdev/platform/sponge/inspection/SpongeInvalidGetterTargetInspection.kt (hunk 3 above)
Comment (RedNesto) [reply src/main/kotlin/com/demonwav/mcdev/platform/sponge/inspection/SpongeInvalidGetterTargetInspection.kt]: It's been a while so I can't remember what I tried to do to end up with this, but I agree this is nonsense, I'll try to keep the same behavior with something better.

**Comment by DenWav** [comment]
Thanks, will see about finishing the review and getting it merged by tomorrow.

**Pull request closed without merging**